File: packages/typescript/src/memory.ts
import {
	type Content,
	type CustomMetadata,
	type DescriptionMetadata,
	type DocumentMetadata,
	type FragmentMetadata,
	type Memory,
	type MemoryMetadata,
	MemoryType,
	type MessageMemory,
	type MessageMetadata,
	type UUID,
} from "./types";

export function createMessageMemory(params: {
	id?: UUID;
	entityId: UUID;
	agentId?: UUID;
	roomId: UUID;
	content: Content & { text: string };
	embedding?: number[];
}): MessageMemory {
	const now = Date.now();
	return {
		...params,
		createdAt: now,
		metadata: {
			type: MemoryType.MESSAGE,
			timestamp: now,
			scope: params.agentId ? "private" : "shared",
		},
	};
}

export function isDocumentMetadata(
	metadata: MemoryMetadata,
): metadata is DocumentMetadata {
	return metadata.type === MemoryType.DOCUMENT;
}

/**
 * Type guard to check if a memory metadata is a FragmentMetadata
 * @param metadata The metadata to check
 * @returns True if the metadata is a FragmentMetadata
 */
export function isFragmentMetadata(
	metadata: MemoryMetadata,
): metadata is FragmentMetadata {
	return metadata.type === MemoryType.FRAGMENT;
}

export function isMessageMetadata(
	metadata: MemoryMetadata,
): metadata is MessageMetadata {
	return metadata.type === MemoryType.MESSAGE;
}

/**
 * Type guard to check if a memory metadata is a DescriptionMetadata
 * @param metadata The metadata to check
 * @returns True if the metadata is a DescriptionMetadata
 */
export function isDescriptionMetadata(
	metadata: MemoryMetadata,
): metadata is DescriptionMetadata {
	return metadata.type === MemoryType.DESCRIPTION;
}

export function isCustomMetadata(
	metadata: MemoryMetadata,
): metadata is CustomMetadata {
	return (
		metadata.type !== MemoryType.DOCUMENT &&
		metadata.type !== MemoryType.FRAGMENT &&
		metadata.type !== MemoryType.MESSAGE &&
		metadata.type !== MemoryType.DESCRIPTION
	);
}

/**
 * Memory type guard for document memories
 */
export function isDocumentMemory(
	memory: Memory,
): memory is Memory & { metadata: DocumentMetadata } {
	return (
		memory.metadata !== undefined &&
		memory.metadata.type === MemoryType.DOCUMENT
	);
}

/**
 * Memory type guard for fragment memories
 */
export function isFragmentMemory(
	memory: Memory,
): memory is Memory & { metadata: FragmentMetadata } {
	return (
		memory.metadata !== undefined &&
		memory.metadata.type === MemoryType.FRAGMENT
	);
}

export function getMemoryText(memory: Memory, defaultValue = ""): string {
	return memory.content.text ?? defaultValue;
}

File: packages/python/elizaos/types/memory.py
from __future__ import annotations

from elizaos.types.generated.eliza.v1 import memory_pb2


class _MemoryTypeValue(str):
    """String that also has a .value property for enum compatibility."""

    @property
    def value(self) -> str:  # noqa: D401
        return str(self)


class MemoryType:
    """Memory type constants (mirrors the TypeScript MemoryType enum)."""

    MESSAGE = _MemoryTypeValue("MESSAGE")
    DOCUMENT = _MemoryTypeValue("DOCUMENT")
    FRAGMENT = _MemoryTypeValue("FRAGMENT")
    DESCRIPTION = _MemoryTypeValue("DESCRIPTION")
    CUSTOM = _MemoryTypeValue("CUSTOM")


BaseMetadata = memory_pb2.BaseMetadata
DocumentMetadata = memory_pb2.DocumentMetadata
FragmentMetadata = memory_pb2.FragmentMetadata
MessageMetadata = memory_pb2.MessageMetadata
DescriptionMetadata = memory_pb2.DescriptionMetadata
CustomMetadata = memory_pb2.CustomMetadata
MemoryMetadata = memory_pb2.MemoryMetadata
Memory = memory_pb2.Memory
MessageMemory = memory_pb2.MessageMemory

__all__ = [
    "MemoryType",
    "BaseMetadata",
    "DocumentMetadata",
    "FragmentMetadata",
    "MessageMetadata",
    "DescriptionMetadata",
    "CustomMetadata",
    "MemoryMetadata",
    "Memory",
    "MessageMemory",
]
